src/nfa/delta.cc

```diff
@@ -230,87 +247,79 @@ Delta::const_iterator& Delta::const_iterator::operator++()
         targets_position = post_iterator->targets.begin();
     }
 
+    transition.src = current_state;
+    transition.symb = post_iterator->symbol;
+    transition.tgt = *targets_position;
+
     return *this;
 }
 
-Delta::const_iterator Delta::const_iterator::operator++(int) {
-    const const_iterator tmp = *this;
+const Delta::transitions_const_iterator Delta::transitions_const_iterator::operator++(int) {
+    const transitions_const_iterator tmp = *this;
     ++(*this);
     return tmp;
 }
 
-Delta::const_iterator& Delta::const_iterator::operator=(const Delta::const_iterator& x) {
+Delta::transitions_const_iterator& Delta::transitions_const_iterator::operator=(const Delta::transitions_const_iterator& x) {
+    // FIXME: this->post is never updated, because it is a const reference to std::vector which does not have assignment
+    //  operator defined.
     this->post_iterator = x.post_iterator;
     this->targets_position = x.targets_position;
     this->current_state = x.current_state;
     this->is_end = x.is_end;
-
+    transition.src = x.transition.src;
+    transition.symb = x.transition.symb;
+    transition.tgt = x.transition.tgt;
     return *this;
 }
 
-bool Mata::Nfa::operator==(const Delta::const_iterator& a, const Delta::const_iterator& b) {
-    if (a.is_end && b.is_end) {
+bool Mata::Nfa::Delta::transitions_const_iterator::operator==(const Delta::transitions_const_iterator& other) const {
+    if (is_end && other.is_end) {
         return true;
-    } else if ((a.is_end && !b.is_end) || (!a.is_end && b.is_end)) {
+    } else if ((is_end && !other.is_end) || (!is_end && other.is_end)) {
         return false;
     } else {
-        return a.current_state == b.current_state && a.post_iterator == b.post_iterator
-               && a.targets_position == b.targets_position;
-    }
-}
-
-State Delta::find_max_state() {
-    size_t max = 0;
-    State src = 0;
-    for (Post & p: posts) {
-        if (src > max)
-            max = src;
-        for (Move & m: p) {
-            if (!m.targets.empty())
-                if (m.targets.back() > max)
-                    max = m.targets.back();
-        }
-        src++;
+        return current_state == other.current_state && post_iterator == other.post_iterator
+               && targets_position == other.targets_position;
     }
-    return max;
 }
 
-std::vector<Post> Delta::transform(const std::function<State(State)>& lambda) const {
-    std::vector<Post> cp_post_vector;
+std::vector<StatePost> Delta::transform(const std::function<State(State)>& lambda) const {
```

Comment:
Maybe rename the `lambda` to `transform_state`?

Reply:
hm, maybe, or maybe something like state_transformer, or other words, like renumber, modify, change
Several other things in fact, not related to this PR:
- the name of the function could rather be tranform_states, not clear what is meant by transform otherwise (e.g. renumber_states would be better in fact ...)
- the comment on monotonicity in delta could be made more fool proof, one doubts about what is meant by monotonic ( it is x<y --> f(x) < f(y), right? )
- can the function work more in-place, without copying the vectors of targets to cp_dest, just modify the existing targets?

Reply:
All good suggestions. We will work on that in future PRs.

Reply:
I changed the name and extended the comment. However, I think that making this function in-place would be useless. One will never call this function to modify the current NFA. What would be the point, if the source states remain the same? The question might be whether we need to have this function in the interface at all, or whether it can be only a local utility function. This function is supposed to behave the same as [`std::transform`](https://en.cppreference.com/w/cpp/algorithm/transform) (hence the original name) and is useful only for creating a new Delta. We could discuss this further.